File: C(++)/029. Divide Two Integers.cpp
/*
029. Divide Two Integers
https://leetcode.com/problems/divide-two-integers/

cpp 6 ms
*/

/*
Divide two integers without using multiplication, division and mod operator.

If it is overflow, return MAX_INT.
*/

class Solution {
#define LL long long
#define INF 0x7fffffff
#define MNF -2147483648
public:
    LL iabs(LL a)
    {
        return a<0?-a:a;
    }
    int divide(int dividend, int divisor) {
        LL a=dividend,b=divisor,sign=1;
        if (a<0&&b>0||a>0&&b<0)
            sign=-1;
        a=iabs(a);
        b=iabs(b);
        if (a<b) return 0;
        if (b==0) return INF;
        LL ret=0;
        while (a>=b)
        {
            LL t=b,d=1;
            while (a>=t)
            {
                ret+=d;
                a-=t;
                t<<=1;
                d<<=1;
            }
        }
        if (ret>INF&&sign==1) return INF;
        if (ret<MNF&&sign==-1) return MNF;
        if (sign==-1) ret-=ret+ret;
        return ret;
    }
};

File: Java/029. Divide Two Integers.java
/*
029. Divide Two Integers
https://leetcode.com/problems/divide-two-integers/

java 39 ms
*/

public class Solution {
	public int divide(int dividend, int divisor) {
		long a=dividend,b=divisor;
		if (b==0||a==Integer.MIN_VALUE&&b==-1)
			return Integer.MAX_VALUE;
		long flag=0,ret=0,r=1;
		if (a<0)
		{
			flag^=1;
			a=-a;
		}
		if (b<0)
		{
			flag^=1;
			b=-b;
		}
		
		long _b=b;
		if (b==1) ret=a;
		else if (b==2) ret=a>>1;
		else
			while (b<=a)
			{
				// 快速减的思路
				while (b<=a)
				{
					ret+=r;
					a-=b;
					r+=r;
					b+=b;
				}
				// b大到一定数会大于余下的a
				// 但此时a可能大于原来的b，所以重复从_b开始
				b=_b;
				r=1;
			}
		if (flag==1) ret=-ret;
		return (int)ret;
	}
}
